pkg/operator/ceph/cluster/osd/spec.go

```diff
@@ -104,17 +105,40 @@ if [[ "$CV_MODE" == "lvm" ]]; then
 	# remove the temporary directory
 	rm --recursive --force "$TMP_DIR"
 else
-	ARGS=(--device ${DEVICE} --no-systemd --no-tmpfs)
-	if [ -n "$METADATA_DEVICE" ]; then
-		ARGS+=(--block.db ${METADATA_DEVICE})
-	fi
-	if [ -n "$WAL_DEVICE" ]; then
-		ARGS+=(--block.wal ${WAL_DEVICE})
+	# 'ceph-volume raw list' (which the osd-prepare job uses to report OSDs on nodes)
+	#  returns user-friendly device names which can change when systems reboot. To
+	# keep OSD pods from crashing repeatedly after a reboot, we need to check if the
+	# block device we have is still correct, and if it isn't correct, we need to
+	# scan all the disks to find the right one.
+	OSD_LIST="$(mktemp)"
+
+	function find_device() {
+		# jq would be preferable, but might be removed for hardened Ceph images
+		# python3 should exist in all containers having Ceph
+		python3 -c "
+import sys, json
+for _, info in json.load(sys.stdin).items():
+	if info['osd_id'] == $OSD_ID:
+		print(info['device'], end='')
+		print('found device: ' + info['device'], file=sys.stderr) # log the disk we found to stderr
+		sys.exit(0)  # don't keep processing once the disk is found
+sys.exit('no disk found with OSD ID $OSD_ID')
+"
+	}
+
+	ceph-volume raw list "$DEVICE" > "$OSD_LIST"
+	cat "$OSD_LIST"
+
+	if ! find_device < "$OSD_LIST"; then
+		ceph-volume raw list > "$OSD_LIST"
+		cat "$OSD_LIST"
+
+		DEVICE="$(find_device < "$OSD_LIST")"
```

Comment:
I think we should error out properly if `DEVICE` is empty instead of letting `ceph-volume` fail. What do you think?

Reply:
That's not a concern. The python script will fail with the `sys.exit(msg)` line if it can't find the device, and `set -o pipefail` will cause the script to fail.